### DISCUSSION TITLE
Assign superset roles to Keycloak authenticated users using the AUTH_ROLES_MAPPING

### DISCUSSION THREAD

**Discussion by benelot** [apache/superset#27902] [General]
I am currently working on integrating superset with keycloak to use the ldap/AD of our company. Integration works fine such that the AD users can login with their password. Through the use of a role-mapper in keycloak and a configuration in client scope, I even get the ldap groups to show up as 'role_keys' as part of keycloak's user_info response. However, it seems like superset does not pick up these roles at all.

I am following the guide of restack (https://www.restack.io/docs/superset-knowledge-apache-superset-keycloak-integration) that works very well so far up to that step.
 From the link:
Step 3: Map Keycloak Roles to Superset Permissions

Use the AUTH_ROLES_MAPPING configuration in Superset to map Keycloak roles to Superset roles. For instance:
```
AUTH_ROLES_MAPPING = { "Superset_Admin": ["Admin"], "Superset_Alpha": ["Alpha"], "Superset_Gamma": ["Gamma"] }
```

I additionally had to find out about how to set the client scope for role_keys. The role_keys were suggested by several users on stackoverflow and also superset's authentication base Flask App Builder suggests that all OAuth Providers just need to provide 'role_keys' in user_info to make role_mapping work (" # however, if you customize userinfo retrieval to include 'role_keys', this will work for other providers" from https://flask-appbuilder.readthedocs.io/en/v4.3.0/security.html#authentication-oauth). I even checked using curl that the user_info really contains a 'role_keys' attribute in the json response. 

What am I doing wrong? Do I really need to override the Apache Security Manager and do the role mapping myself? I expected that if I followed the requirements by flask, I would get the mapping out of the box.

I am happy to provide a full recipe for the integration once I get the roles mapped on login.

I really appreciate any help!

Thx,

**Comment by blarghmatey** [comment]
The missing piece is that the out-of-the-box support for Keycloak omits the `role_keys` attribute from the return value in Flask AppBuilder (you can see that [here](https://github.com/dpgaspar/Flask-AppBuilder/blob/c65e067f09e741c00322221263c8599b8e8811d5/flask_appbuilder/security/manager.py#L693-L698))

To address that shortcoming you'll need to create and use a custom security manager class to manage passing that `role_keys` attribute through. For example, [this](https://github.com/mitodl/ol-infrastructure/blob/main/src/ol_superset/pythonpath/superset_config.py#L88-L103) is what I did in my deployment to handle that case.

**Comment by benelot** [comment]
Thank you for the perfect answer! I think I am going to pull-request to Flask-AppBuilder to just include the `role_keys` there as well. They write that by providing the 'role_keys', it should pick up roles for other providers, and including them in the default keycloak call sounds like an improvement toward that promise.

**Reply by myselfuser1** [reply]
@blarghmatey Can you please help me as well, i need to fetch my keycloak roles, i tried the solution which you gave me but every time i sign in , the public role is assigned to the user on superset irrespective of what role is present for that user in keycloak, here is the code of `superset_config.py` file


```
SECRET_KEY = 'my_secret_key'
AUTH_TYPE = AUTH_OAUTH
LOGOUT_REDIRECT_URL='https://localhost:3000/auth/realms/master/protocol/openid-connect/logout'
AUTH_USER_REGISTRATION = True
# AUTH_USER_REGISTRATION_ROLE = "Gamma"
OAUTH_PROVIDERS = [
    {
        'name': 'keycloak',
        'token_key': 'access_token',
        'icon': 'fa-key',
        'remote_app': {
            'client_id': 'MY_CLIENT_ID',  # Replace with your Keycloak client ID
            'client_secret': 'MY_CLIENT_SECRET',  # Replace with your Keycloak client secret
            'api_base_url': 'https://localhost:3000/auth/realms/master/protocol/openid-connect/',  # Replace with your Keycloak realm URL
            'client_kwargs': {
                'scope': 'openid profile email',
                'roles_key': 'realm_access.roles',
            },
            'server_metadata_url': 'https://localhost:3000/realms/master/.well-known/openid-configuration',
            'api_base_url': 'https://localhost:3000/realms/master/protocol/',
            'access_token_url': 'https://localhost:3000/auth/realms/master/protocol/openid-connect/token',  # Replace with your Keycloak token URL
            'authorize_url': 'https://localhost:3000/auth/realms/master/protocol/openid-connect/auth',  # Replace with your Keycloak authorization URL
            'jwks_uri': 'https://localhost:3000/auth/realms/master/protocol/openid-connect/certs',  # Replace with your Keycloak JWKS URI
        },
    }
]
AUTH_ROLES_SYNC_AT_LOGIN = True

JWT_ALGORITHM = "RS256"
# URL to the public key endpoint
public_key_url = "https://localhost:3000/auth/realms/master"


def fetch_keycloak_rs256_public_cert():
    with urllib.request.urlopen(public_key_url) as response:  # noqa: S310
        public_key_url_response = json.load(response)
    public_key = public_key_url_response["public_key"]
    if public_key:
        pem_lines = [
            "-----BEGIN PUBLIC KEY-----",
            public_key,
            "-----END PUBLIC KEY-----",
        ]
        cert_pem = "\n".join(pem_lines)
    else:
        cert_pem = "No cert found"
    return cert_pem

print("aaaaa")
JWT_PUBLIC_KEY = fetch_keycloak_rs256_public_cert()
print("bbbbbb")
AUTH_ROLES_MAPPING = {
    'Super Admin': ['Admin', 'sql_lab'],
    'Portal Admin': ['Admin',],
    'Company Admin': ['Admin',],
    'Admin': ['Admin',],
    'admin': ['Admin',],
    'Non C-Cash User': ['Public'],
    'Ex C-Cash User': ['Public'],
    'External Users': ['Public'],
    'Licensed C-Cash User': ['Alpha'],
    'C-Cash User': ['Gamma'],
}
AUTH_ROLES_SYNC_AT_LOGIN = True
AUTH_USER_REGISTRATION = True
AUTH_USER_REGISTRATION_ROLE = "Public"

class CustomSsoSecurityManager(SupersetSecurityManager):
    def oauth_user_info(self, provider, response=None):  # noqa: ARG002
        me = self.appbuilder.sm.oauth_remotes[provider].get("openid-connect/userinfo")
        me.raise_for_status()
        data = me.json()
        logging.debug("User info from Keycloak: %s", data)
        return {
            "username": data.get("preferred_username", ""),
            "first_name": data.get("given_name", ""),
            "last_name": data.get("family_name", ""),
            "email": data.get("email", ""),
            "role_keys": data.get("role_keys", []),
        }

    def load_user_jwt(self, _jwt_header, jwt_data):
        username = jwt_data["preferred_username"]
        user = self.find_user(username=username)
        if user.is_active:
            # Set flask g.user to JWT user, we can't do it on before request
            g.user = user
            return user
        return None

CUSTOM_SECURITY_MANAGER = CustomSsoSecurityManager
```


what else am i missing?

**Comment by myselfuser1** [comment]
Further debugging when i print `data = me.json()` i get this

```
{'sub': '182345b0-c8a1-1233-2da2-c22afe2b882c', 'email_verified': True, 
    'name': 'Abc def', 'preferred_username': 'abc@gmail.com',
 'given_name': 'Abc', 'family_name': 'def', 'email': 'abc@gmail.com'}
```

there is no `role_keys`

**Reply by blarghmatey** [reply]
In Keycloak you need to explicitly add the `role_keys` mapping in the OAuth client definition to ensure that it gets passed through to Superset.

**Reply by chiarapascucci** [reply]
hey @blarghmatey - do you know how to add this mapping in the client definition? Is this something you need to do through KC or are you referring to the superset config code?

**Reply by BraveEvidence** [reply]
@chiarapascucci this helped me https://www.youtube.com/watch?v=psow7KZv7Q8

**Comment by dridk** [comment]
in keycloak, check "add to userinfo" in client scopes > client scope details > mapper details of user realm role. 
In this way, you will get role information in userinfo
